### PR TITLE
PyTorch AMD Build Scripts

### PR DESCRIPTION
**Pull request by Jorghi12**
The PyTorch script files necessary for building on AMD GPUs.

### CODE REVIEW

**Comment by ezyang** [comment]
Hi @Jorghi12, does this mean we are ready to setup AMD builders on CI?

Also, directory structure bikeshedding: can we put this inside `tools`, e.g., as `tools/amd`?

File: amd_build/build_pytorch_amd.py
Comment (ezyang) [inline amd_build/build_pytorch_amd.py]: It would be more robust to use `__file__` to determine where the script lives; otherwise this script is sensitive to the particular directory you run it in.

File: amd_build/build_pytorch_amd.py
Comment (ezyang) [inline amd_build/build_pytorch_amd.py]: I don't see updates to `.gitignore`. Is the copied source tree ignored appropriately by git?

File: amd_build/build_pytorch_amd.py
Comment (ezyang) [inline amd_build/build_pytorch_amd.py]: I'm OK with renaming these files in the native directory ourselves and then removing the special case here.

File: amd_build/build_pytorch_amd.py
Comment (ezyang) [inline amd_build/build_pytorch_amd.py]: Most people aren't going to have this script. How do I get it?

File: amd_build/hip_files/ATen/CMakeLists.txt.hip
Comment (ezyang) [inline amd_build/hip_files/ATen/CMakeLists.txt.hip]: Is this something like, the original `CMakeLists.txt` was copied and then edited until it worked with HIP? What's our ongoing maintenance strategy for noting changes to CMakeLists.txt and getting them to HIP? Would it make sense to try to reintegrate these into the real CMakeLists.txt and having search for HIP be controlled by a flag?

File: amd_build/hip_files/THC/THCApply.cuh.hip
Comment (ezyang) [inline amd_build/hip_files/THC/THCApply.cuh.hip]: Same here: it would be good to know (via a comment at the top of the file), what the substantive differences of this file and `THCApply.cuh` are.

**Comment by ezyang** [comment]
Looking at the patch, it seems sufficiently big that the next time someone merges a change to anything in THC, it will immediately stop working.

In general, I am wondering, would it make more sense to maintain this as a branch, until we can mainline the source code changes so that they work for AMD as well as CUDA? (I was not privy to any of the original discussions, so I apologize if this is something you've already looked at.)

File: amd_build/build_pytorch_amd.py
Comment (Jorghi12) [reply amd_build/build_pytorch_amd.py]: It's going to be on the HIP Codebase here: https://github.com/ROCm-Developer-Tools/pyHIPIFY, obtained from installing HIP/ROCm packages.

File: amd_build/hip_files/ATen/CMakeLists.txt.hip
Comment (Jorghi12) [reply amd_build/hip_files/ATen/CMakeLists.txt.hip]: The CMakeLists.txt files are extremely similar and they already contain a special WITH_ROCM flag (so does the patched setup.py file). They could both be integrated accordingly.

File: amd_build/build_pytorch_amd.py
Comment (ezyang) [reply amd_build/build_pytorch_amd.py]: OK, don't tell me, put it in the code so other people can find it :)

**Comment by Jorghi12** [comment]
Integrated building for HIP into the aten/CMakeLists.txt, aten/src/ATen/CMakeLists.txt, setup.py, and tools/build_pytorch_libs.sh scripts, using the WITH_ROCM flag. And divided the giant patch file into separate individual patches per file in the codebase.

@ezyang - Plan is to merge into master and over time reduce the number of patch files required by making changes to the HIP code base or Hipify Script.

File: aten/CMakeLists.txt
Comment (ezyang) [inline aten/CMakeLists.txt]: This SET() is being applied unconditionally, even if you're not using the ROCM stack.

File: aten/CMakeLists.txt
Comment (ezyang) [inline aten/CMakeLists.txt]: I don't understand this conditional; does ROCM not support C++11?

File: aten/CMakeLists.txt
Comment (ezyang) [inline aten/CMakeLists.txt]: Also confused about this conditional. Does ROCM not support SSE3?

File: aten/CMakeLists.txt
Comment (ezyang) [inline aten/CMakeLists.txt]: I really want to see a message here saying that ROCM is eanbled

File: aten/cmake/FindHIP.cmake
Comment (ezyang) [inline aten/cmake/FindHIP.cmake]: Generally, the cmake file which is actually invoked by `find_package` is given the name `Find*.cmake`. This file is named like this but doesn't follow that convention. It would be clearer to name it something like `LoadHIP.cmake`; just don't use the Find convention.

File: aten/src/ATen/CMakeLists.txt
Comment (ezyang) [inline aten/src/ATen/CMakeLists.txt]: And now HIP seems to support C++11?

File: aten/src/ATen/CMakeLists.txt
Comment (ezyang) [inline aten/src/ATen/CMakeLists.txt]: And now it doesn't support C++11?

File: aten/src/ATen/CMakeLists.txt
Comment (ezyang) [inline aten/src/ATen/CMakeLists.txt]: This looks like it is being unconditionally searched for. You didn't REQUIRE it, but wouldn't it be safer to only search if WITH_ROCM?

File: setup.py
Comment (ezyang) [inline setup.py]: These look like debug prints

**Comment by ezyang** [comment]
OK. @Jorghi12 can you modify `.travis.yml` to add another build configuration where it applies all of the ROCM diffs, and fails if they don't apply cleanly? You can skip this if we're supposed to be getting the AMD CI online in the very short future.

FYI, I think a decent number of patches are going to start tripping over AMD specific code and/or make these patches stop applying. We're going to need clear guidance about what to do in this situation (whether or not it's blocking the PR until you get around to it, or educating developers enough about the AMD stack so they can fix it themselves, or temporarily disabling functionality if it gets out of sync.)

File: setup.py
Comment (apaszke) [inline setup.py]: Why? A hint in the comment would be useful

File: setup.py
Comment (apaszke) [inline setup.py]: Why does `ROCM` affect this one? It's completely unrelated to any GPU stuff

File: tools/amd_build/patches/a_aten_src_ATen_Error.h.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_ATen_Error.h.patch]: Looks like this will break the Windows build. Why can't you make it `!defined(_WIN32) && !defined(__HIP_PLATFORM_HCC__)`?

File: tools/amd_build/patches/a_aten_src_ATen_cuda_CUDAHalf.cu.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_ATen_cuda_CUDAHalf.cu.patch]: If this file is completely disabled in the HIP build, then maybe it's just simpler to exclude it from the build?

File: tools/amd_build/patches/a_aten_src_ATen_native_cuda_Distributions.cu.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_ATen_native_cuda_Distributions.cu.patch]: This looks really dangerous, because it will basically silently return an uninitialized tensor if you compile with HIP

File: tools/amd_build/patches/a_aten_src_ATen_native_cuda_Embedding.cu.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_ATen_native_cuda_Embedding.cu.patch]: Those functions should be templated on the input type (like the whole kernel is), but have hardcoded `float`

File: tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch]: This also looks like it will silently start returning incorrect results. Would be better to replace `// whitespace` with `throw ...`

File: tools/amd_build/patches/a_aten_src_THCUNN_BCECriterion.cu.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_THCUNN_BCECriterion.cu.patch]: This code looks like it's what's originally been there, except for the `__host__` addition. We could just patch the original file if that helps.

File: tools/amd_build/patches/a_aten_src_THCUNN_SoftMaxCommon.cuh.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_THCUNN_SoftMaxCommon.cuh.patch]: Maybe set this to something larger? 8 blocks is not much

File: tools/amd_build/patches/a_aten_src_THC_THCAllocator.c.patch
Comment (apaszke) [inline tools/amd_build/patches/a_aten_src_THC_THCAllocator.c.patch]: This should also be replaced with `throw ...` instead of silently saying everything went file

File: tools/amd_build/patches/a_torch_csrc_generic_StorageSharing.cpp.patch
Comment (apaszke) [inline tools/amd_build/patches/a_torch_csrc_generic_StorageSharing.cpp.patch]: But why? Those shouldn't matter for HIP. This file should be still built using the regular C++ compiler

File: tools/amd_build/patches/a_torch_csrc_utils_python_strings.h.patch
Comment (apaszke) [inline tools/amd_build/patches/a_torch_csrc_utils_python_strings.h.patch]: Same here

**Review by apaszke — COMMENTED** [review]
Nice work @Jorghi12! I do also have a few concerns about our current strategy, and the contents of the patches:
1. Why did we have to change so much of code that's completely unrelated to GPU? A few examples:
  - TH changes
  - The `docs/` directory
  - Many docstrings in Python files
2. There are some things that I don't quite understand:
  - Why do files in the `cudnn/` directory in ATen gained an `CuDNN` suffix? It's clear from the directory name and seems unnecessary. That's not a patch, so it will change how the repo looks like.
  - Why do we change Caffe2? We probably don't want to patch that.
3. Why do we need those `.hip` files. They look as if they were pretty much copied from THC, so why is it better to store them like this instead of patches? Some of them contain pretty complicated platform independent logic (think all the stride tricks in `THCApply`), so I'd rather avoid duplicating them.
4. I have very similar concerns to @ezyang. How long do we plan to keep the codebase in this state?

**Comment by apaszke** [comment]
Hmm what are the files that have the same file names and cause the conflicts? I can't see them.

Great stuff @Jorghi12 looking forward to more simplifications! Just FYI I think CUDA 9 also treats half as a pretty much first-class type, but we can't rely on that if we want to support CUDA 8.

**Comment by wsttiger** [comment]
Hi @apaszke we're not modifying the original source, so the CUDA 8 (and below) half should be fine in THC.

**Comment by wsttiger** [comment]
Having said that, perhaps this would be a good time add cross-platform support for __fp16 / half in THC.

**Comment by apaszke** [comment]
@wsttiger oh yes, I know that. I only meant it as a general note. What do you mean by cross platform support for half?

**Comment by Jorghi12** [comment]
@apaszke Yeah you're right, the old BatchNorm.cpp isn't here anymore. The remaining duplicate is Conv.cpp which is [here](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/mkldnn/Conv.cpp). To make things cleaner, I'll rename the Conv to ConvCuDNN using the preprocessor script (amd_build/build_pytorch_amd.py) and keep the original PyTorch file names.

Regarding the CUDA8/CUDA9, I see that makes sense.

**Comment by ezyang** [comment]
@apaszke I still stand by my earlier suggestion. I don't see that there's anything wrong with naming the cudnn convolution CuDNNConv.cpp and the MKLDNN convolution MKLDNNConv.cpp. It seems like a reasonable enough bikeshed color to me, and if it simplifies the build I'm all for it.

**Comment by apaszke** [comment]
@ezyang sorry, which suggestion do you have in mind?

**Comment by ezyang** [comment]
network flaked on OS X failure

**Comment by wsttiger** [comment]
@apaszke I think that @Jorghi12 has implemented __half_raw which should be a cleaner implementation (then whatever we were doing before:))

File: setup.py
Comment (Jorghi12) [reply setup.py]: On the ROCm stack, the HCC compiler generates both the host & device code. I briefly spoke with a few developers at AMD about allowing other compilers such as GCC to be used on host - though it's not on their road map. It'll require modifications to [HIPCC](https://github.com/ROCm-Developer-Tools/HIP/blob/master/bin/hipcc). I'll give it a dive soon.

(HIPCC is what we set as the compiler in CMakeLists.txt)

File: amd_build/build_pytorch_amd.py
Comment (csuji) [reply amd_build/build_pytorch_amd.py]: @Jorghi12 Hi, I am trying to build your pytorch sending_pr branch for AMD but it seams the current pyHIPIFY is not up to date, option --include-dirs missing and I get the following error:
```
Total number of replaced kernel launches: 1433
Traceback (most recent call last):
  File "/opt/rocm/bin/hipify-python.py", line 845, in <module>
    main()
  File "/opt/rocm/bin/hipify-python.py", line 809, in main
    txt = disable_function(txt, func, 1)
  File "/opt/rocm/bin/hipify-python.py", line 329, in disable_function
    "return_type": the_match.group(2).strip(),
AttributeError: 'NoneType' object has no attribute 'group'
```
and is it ok to get the following error while patching?:
```
error: patch failed: aten/src/THC/THCTensorRandom.cpp:83
error: aten/src/THC/THCTensorRandom.cpp: patch does not apply
```

File: aten/CMakeLists.txt
Comment (ezyang) [inline aten/CMakeLists.txt]: Note for a follow up PR: we're moving all of the cmake modules into the top-level cmake directory, so LoadHIP should live there similarly.

File: tools/amd_build/patches/a_aten_src_ATen_ATen.h.patch
Comment (ezyang) [inline tools/amd_build/patches/a_aten_src_ATen_ATen.h.patch]: This include runs counter to ATen header philosophy; we shouldn't include CUDA or HIP headers from the user visible ATen/ATen.h header.

File: tools/amd_build/patches/a_aten_src_THC_THCTensorMathReduce.cuh.patch
Comment (ezyang) [inline tools/amd_build/patches/a_aten_src_THC_THCTensorMathReduce.cuh.patch]: This int cast looks very questionable

File: tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch
Comment (ezyang) [inline tools/amd_build/patches/a_aten_src_THCS_generic_THCSTensorMath.cu.patch]: No op?

File: tools/amd_build/patches/a_aten_src_THCUNN_IndexLinear.cu.patch
Comment (ezyang) [inline tools/amd_build/patches/a_aten_src_THCUNN_IndexLinear.cu.patch]: Might be simplest to just unconditionally make all of these non-static.

Comment (ezyang) [reply tools/amd_build/patches/a_aten_src_THCUNN_IndexLinear.cu.patch]: Note about HIP limitations w.r.t. static kernels would be helpful, somewhere.

File: tools/amd_build/patches/a_aten_src_THCUNN_generic_SpatialCrossMapLRN.cu.patch
Comment (ezyang) [inline tools/amd_build/patches/a_aten_src_THCUNN_generic_SpatialCrossMapLRN.cu.patch]: So, HIP version of this just doesn't work, is that right? Wouldn't it be better to raise an error than just skip the kernel?

**Comment by ezyang** [comment]
Thank you @Jorghi12, the new patches look much much improved! Just to summarize the things we're planning to fix later:

- Get AMD builders online. (Can we get more than one box from AMD, so that we're not perpetually low on capacity?)
- Eliminate more of the diffs
  - Don't `#include "hip/hip_runtime.h"` from user-facing headers
  - `&& !defined(__HIP_PLATFORM_HCC__)` tests don't have to be patched, we can unconditionally include them in our code
  - `WITH_ROCM` conditionals in cmake can be put straight in the cmake
  - We should probably make a `cuda_based_add_executable` which either switches to `cuda_add_executable` or `hip_add_executable`
  - Ifdef the bitlanes file
- Fix some correctness issues
  - Stop specializing abs/pow to float

Also, longer term thing I am thinking about: with https://github.com/pytorch/pytorch/pull/7275 coming down the pipe, it would be nice to think of a way where you can easily build CPU/CUDA/AMD all in one go into separate libraries, not controlled by a build flag.

File: tools/amd_build/patches/a_aten_src_THC_THCTensorMathReduce.cuh.patch
Comment (Jorghi12) [reply tools/amd_build/patches/a_aten_src_THC_THCTensorMathReduce.cuh.patch]: @wsttiger can explain further on this. There are issues with the compiler that we needed to work around.

**Review by apaszke — COMMENTED** [review]
Looks great!

Is THCTensorRandom.cu.hip really so much different than THCTensorRandom.cu? Why isn't it a patch.

**Comment by Jorghi12** [comment]
Updates:
@apaszke I removed the hip files completely and left THCTensorRandom.cu in a minimal patch form. (Static casts are added due to the annoying implementation of hipKernelLaunch.)

Top level CMake Changes:
- The LoadHIP.cmake script has been added to cmake/public since it'll be useful for Caffe2 build as well.
- The command torch_cuda_based_add_executable was added inside of the cmake/public/utils.cmake. Similarly done for torch_cuda_based_add_library.

File: setup.py
Comment (apaszke) [inline setup.py]: Can you please remove this line?

**Review by apaszke — APPROVED** [review]
LGTM. One minor nit. Should be ready to merge once @ezyang approves the CMake part.

**Comment by Jorghi12** [comment]
The build for pr/pytorch-linux-xenial-py3-clang5-asan was hanging for 3 hours previously so I restarted the build on Jenkins: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-py3-clang5-asan-build/4652/console

**Pull request merged**
